kedro/config/omegaconf_config.py

```diff
@@ -0,0 +1,258 @@
+
+        config.update(env_config)
+
+        if not config:
+            raise MissingConfigException(
+                f"No files of YAML or JSON format found in {base_path} or {env_path} matching"
+                f" the glob pattern(s): {[*self.config_patterns[key]]}"
+            )
+        return config
+
+    def __repr__(self):  # pragma: no cover
+        return (
+            f"OmegaConfLoader(conf_source={self.conf_source}, env={self.env}, "
+            f"config_patterns={self.config_patterns})"
+        )
+
+    def load_and_merge_dir_config(self, conf_path: str, patterns: Iterable[str]):
+        """Recursively load and merge all configuration files in a directory using OmegaConf,
+        which satisfy a given list of glob patterns from a specific path.
+
+        Args:
+            conf_path: Path to configuration directory.
+            patterns: List of glob patterns to match the filenames against.
+
+        Raises:
+            MissingConfigException: If configuration path doesn't exist or isn't valid.
+            ValueError: If two or more configuration files contain the same key(s).
+            ParserError: If config file contains invalid YAML or JSON syntax.
+
+        Returns:
+            Resulting configuration dictionary.
+
+        """
+        if not Path(conf_path).is_dir():
+            raise MissingConfigException(
+                f"Given configuration path either does not exist "
+                f"or is not a valid directory: {conf_path}"
+            )
+
+        paths = [
+            Path(each).resolve()
+            for pattern in patterns
+            for each in iglob(f"{str(conf_path)}/{pattern}", recursive=True)
+        ]
+        deduplicated_paths = set(paths)
+        config_files_filtered = [
+            path for path in deduplicated_paths if self._is_valid_config_path(path)
+        ]
+
+        config = {}
+        aggregate_config = []
+        seen_file_to_keys = {}  # type: Dict[Path, AbstractSet[str]]
+
+        for config_filepath in config_files_filtered:
+            try:
+                single_config = OmegaConf.load(config_filepath)
+                config = single_config
+            except (ParserError, ScannerError) as exc:
+                assert exc.problem_mark is not None
+                line = exc.problem_mark.line
+                cursor = exc.problem_mark.column
+                raise ParserError(
+                    f"Invalid YAML or JSON file {config_filepath}, unable to read line {line}, "
+                    f"position {cursor}."
+                ) from exc
+
+            _check_duplicate_keys(
+                seen_file_to_keys, config_filepath, dict(single_config)
+            )
+            seen_file_to_keys[config_filepath] = single_config.keys()
+            aggregate_config.append(single_config)
+
+        if len(aggregate_config) > 1:
+            return dict(OmegaConf.merge(*aggregate_config))
+
+        return config
```

Comment:
Can we remove the special handling of `single_config` when `len(aggregate_config) == 1`? It should handle with len == 1 and thus no need to have an extra variable `config` and
```suggestion
        single_config = {}
        aggregate_config = []
        seen_file_to_keys = {}  # type: Dict[Path, AbstractSet[str]]

        for config_filepath in config_files_filtered:
            try:
                single_config = OmegaConf.load(config_filepath)
            except (ParserError, ScannerError) as exc:
                assert exc.problem_mark is not None
                line = exc.problem_mark.line
                cursor = exc.problem_mark.column
                raise ParserError(
                    f"Invalid YAML or JSON file {config_filepath}, unable to read line {line}, "
                    f"position {cursor}."
                ) from exc

            _check_duplicate_keys(
                seen_file_to_keys, config_filepath, dict(single_config)
            )
            seen_file_to_keys[config_filepath] = single_config.keys()
            aggregate_config.append(single_config)


        return dict(OmegaConf.merge(*aggregate_config))

```